src/lib/mina_base/parties_logic.ml

```diff
@@ -601,25 +592,25 @@ module Make (Inputs : Inputs_intf) = struct
       }
     in
     let global_excess_update_failed = ref Bool.true_ in
-    let global_state, local_state =
-      ( h.perform
-          (Modify_global_excess
-             ( global_state
-             , fun amt ->
-                 let res, `Overflow overflow =
-                   Amount.Signed.add_flagged amt
-                     (Amount.Signed.of_unsigned local_state.excess)
-                 in
-                 (global_excess_update_failed :=
-                    Bool.(!update_global_state &&& overflow)) ;
-                 (update_global_state :=
-                    Bool.(!update_global_state &&& not overflow)) ;
-                 Amount.Signed.if_ !update_global_state ~then_:res ~else_:amt ))
-      , { local_state with
-          excess =
-            Amount.if_ update_local_excess ~then_:Amount.zero
-              ~else_:local_state.excess
-        } )
+    let global_state =
+      let amt = Global_state.fee_excess global_state in
+      let res, `Overflow overflow =
+        Amount.Signed.add_flagged amt
+          (Amount.Signed.of_unsigned local_state.excess)
+      in
+      (global_excess_update_failed := Bool.(!update_global_state &&& overflow)) ;
```

Comment:
Can these ref cell updates be avoided by creating new let-bindings? That is, return these values here, in addition to `global_state`.

Reply:
Good idea, done.